main.py

```diff
@@ -85,8 +94,145 @@ def recording_file(rec_file):
+    """
+    Loads the data from a given data directory, giving back both inputs and labels
+    :param data_dir: The directory to load data from
+    :param preprocess_method: The method to use for preprocessing (mfcc | fft)
+    :return: The inputs and labels from the data directory
+    """
+    inputs = None
+    labels = None
+
+    accent_class_folders = [folder for folder in os.listdir(data_dir)
+                            if os.path.isdir(os.path.join(data_dir, folder))]
+    for folder in accent_class_folders:
+        data_file = os.path.join(data_dir, folder, f"{folder}-{preprocess_method}.npy")
+        class_data = Io.read_audio_data(data_file)
+        class_labels = np.full((class_data.shape[0]), model.accent_classes.index(folder))
+
+        if inputs and labels:
+            inputs = np.concatenate(inputs, class_data)
+            labels = np.concatenate(labels, class_labels)
+        else:
+            inputs = class_data
+            labels = class_labels
+
+    return inputs, labels
+
+
+def train(model, epochs, train_data_dir, save_file=None, preprocess_method="mfcc"):
+    """
+    Trains the model on the given training data, checkpointing the weights to the given file
+    after every epoch.
+    :param model: The model to train.
+    :param epochs: Number of epochs to train for.
+    :param train_data_dir: A directory of the training data to use
+    :param save_file: The file to save the model weights to.
+    :param preprocess_method: The preprocess method to use for the inputs to the model (mfcc | fft)
+    :return: The trained model
+    """
+
+    train_inputs, train_labels = get_data_from_dir(train_data_dir, preprocess_method)
+
+    assert train_inputs is not None
+    assert train_labels is not None
+    assert train_inputs.shape[0] == train_labels.shape[0]
+    dataset_size = train_labels.shape[0]
+
+    for e in range(epochs):
+
+        # Shuffle the dataset before each epoch
+        new_order = np.random.permutation(dataset_size)
+        train_inputs = train_inputs[new_order]
+        train_labels = train_labels[new_order]
+
+        # Run training in batches
+        for batch_start in range(0, dataset_size, model.batch_size):
+            batch_inputs = train_inputs[batch_start:batch_start + model.batch_size]
+            batch_labels = train_labels[batch_start:batch_start + model.batch_size]
+
+            with tf.GradientTape() as tape:
+                loss = model.loss(batch_inputs, batch_labels)
+
+            grads = tape.gradient(loss, model.trainable_variables)
+            model.optimizer.apply_gradients(zip(grads, model.trainable_variables))
+
+        # Print loss and accuracy
+        epoch_loss = model.loss(train_inputs, train_labels)
+        epoch_acc = model.accuracy(train_inputs, train_labels)
+        print(f"Epoch {e}/{epochs} | Loss: {epoch_loss} | Accuracy: {epoch_acc}")
+
+        # Save the model at the end of the epoch
+        if save_file:
```

Comment:
Right now we're checkpointing at the end of every epoch. Depending on training speed, maybe we should checkpoint every x batches instead.

Reply:
Yeah, I'll look into it if training is super slow